### PR TITLE
[SPARK-49877][SQL] Change classifyException function signature: add isRuntime argument

### PR DESCRIPTION
**Pull request by ivanjevtic-db**
### What changes were proposed in this pull request?
The proposal is to update the classifyException function so that it can return either ```AnalysisException``` or ```SparkRuntimeException```. This is achieved by adding a new parameter, ```isRuntime```, and modifying the return type to be ```Throwable with SparkThrowable``` for compatibility with both types.

### Why are the changes needed?
The changes are needed to allow the classifyException function to be used in execution part of the code.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Not needed.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 1]
```diff
@@ -741,16 +741,30 @@ abstract class JdbcDialect extends Serializable with Logging {
    * @param errorClass The error class assigned in the case of an unclassified `e`
    * @param messageParameters The message parameters of `errorClass`
    * @param description The error description
-   * @return `AnalysisException` or its sub-class.
+   * @param isRuntime Whether the exception is a runtime exception or not.
+   * @return `SparkThrowable` or its sub-class.
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:745]: SparkThrowable + Throwable

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 2]
```diff
@@ -741,16 +741,30 @@ abstract class JdbcDialect extends Serializable with Logging {
...
   }
 
+  @deprecated("Please override the classifyException method with `isRuntime`", "4.0.0")
+  def classifyException(
+      e: Throwable,
+      errorClass: String,
+      messageParameters: Map[String, String],
+      description: String): AnalysisException = {
+    new AnalysisException(
+      errorClass = "...",
```
[the remark this answers is not in the document]
Comment (ivanjevtic-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:763]: but we're removing it?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 3]
```diff
@@ -741,13 +741,15 @@ abstract class JdbcDialect extends Serializable with Logging {
...
+   * @param isRuntime Whether the exception is a runtime exception or not.
+   * @return `SparkThrowable + Throwable` or its sub-class.
    */
   def classifyException(
       e: Throwable,
       errorClass: String,
       messageParameters: Map[String, String],
-      description: String): AnalysisException = {
+      description: String,
+      isRuntime: Boolean): Throwable with SparkThrowable = {
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:752]: Sorry for back and forth, but do we need this flag now?

Comment (ivanjevtic-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:752]: We will need it for this pr: #48182(after this is merged I will do some modifications).

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:752]: sgtm

**Comment by ivanjevtic-db** [comment]
The tests are now passing. Could you make a review @MaxGekk? Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/DB2Dialect.scala [hunk 4]
```diff
@@ -154,11 +153,13 @@ private case class DB2Dialect() extends JdbcDialect with SQLConfHelper with NoLe
     s"COMMENT ON SCHEMA ${quoteIdentifier(schema)} IS ''"
   }
 
+
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/DB2Dialect.scala:156]: it is unnecessary changes:
```suggestion

```

**Review by MaxGekk — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 5]
```diff
@@ -741,13 +741,15 @@ abstract class JdbcDialect extends Serializable with Logging {
    * @param errorClass The error class assigned in the case of an unclassified `e`
    * @param messageParameters The message parameters of `errorClass`
    * @param description The error description
-   * @return `AnalysisException` or its sub-class.
+   * @param isRuntime Whether the exception is a runtime exception or not.
+   * @return `SparkThrowable + Throwable` or its sub-class.
    */
   def classifyException(
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:747]: This method has not been released yet, so, we can modify it.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @ivanjevtic-db.
